File: Tests/SentryTests/SentryCrash/TestSentryCrashWrapper.m
#import "TestSentryCrashWrapper.h"
#import "SentryCrash.h"
#import <Foundation/Foundation.h>

@implementation TestSentryCrashWrapper

+ (instancetype)sharedInstance
{
    TestSentryCrashWrapper *instance = [[self alloc] init];
    instance.internalActiveDurationSinceLastCrash = NO;
    instance.internalDurationFromCrashStateInitToLastCrash = 0;
    instance.internalActiveDurationSinceLastCrash = 0;
    instance.internalIsBeingTraced = NO;
    instance.internalIsSimulatorBuild = NO;
    instance.internalIsApplicationInForeground = YES;
    instance.installAsyncHooksCalled = NO;
    instance.uninstallAsyncHooksCalled = NO;
    instance.internalFreeMemorySize = 0;
    instance.internalAppMemorySize = 0;
    instance.internalFreeStorageSize = 0;
    return instance;
}

- (void)startBinaryImageCache
{
    _binaryCacheStarted = YES;
    [super startBinaryImageCache];
}

- (void)stopBinaryImageCache
{
    [super stopBinaryImageCache];
    _binaryCacheStopped = YES;
}

- (BOOL)crashedLastLaunch
{
    return self.internalCrashedLastLaunch;
}

- (NSTimeInterval)durationFromCrashStateInitToLastCrash
{
    return self.internalDurationFromCrashStateInitToLastCrash;
}

- (NSTimeInterval)activeDurationSinceLastCrash
{
    return self.internalActiveDurationSinceLastCrash;
}

- (BOOL)isBeingTraced
{
    return self.internalIsBeingTraced;
}

- (BOOL)isSimulatorBuild
{
    return self.internalIsSimulatorBuild;
}

- (BOOL)isApplicationInForeground
{
    return self.internalIsApplicationInForeground;
}

- (void)installAsyncHooks
{
    self.installAsyncHooksCalled = YES;
}

- (void)uninstallAsyncHooks
{
    self.uninstallAsyncHooksCalled = YES;
}

- (NSDictionary *)systemInfo
{
    return @{};
}

- (bytes)freeMemorySize
{
    return self.internalFreeMemorySize;
}

- (bytes)appMemorySize
{
    return self.internalAppMemorySize;
}

- (bytes)freeStorageSize
{
    return self.internalFreeStorageSize;
}

@end

File: Tests/SentryTests/SentryCrash/TestSentryCrashWrapper.swift
@_spi(Private) @testable import Sentry
import Foundation

class TestSentryCrashReporterState: NSObject, SentryCrashReporterState {
    var internalInstalled = false
    var internalCrashedLastLaunch = false

    var installed: Bool { internalInstalled }
    var crashedLastLaunch: Bool { internalCrashedLastLaunch }
}

#if !SDK_V10
/// V9 test double for the broad SentryCrash reporter adapter.
final class TestSentryCrashReporter: TestSentryCrashReporterState, SentryCrashReporter {
    var internalDurationFromCrashStateInitToLastCrash: TimeInterval = 0
    var internalActiveDurationSinceLastCrash: TimeInterval = 0
    var internalIsSimulatorBuild = false
    var internalIntrospectMemory: Bool = true

    convenience init(processInfoWrapper: SentryProcessInfoSource) {
        self.init()
    }

    var durationFromCrashStateInitToLastCrash: TimeInterval { internalDurationFromCrashStateInitToLastCrash }
    var activeDurationSinceLastCrash: TimeInterval { internalActiveDurationSinceLastCrash }
    var isSimulatorBuild: Bool { internalIsSimulatorBuild }
    var introspectMemory: Bool {
        get { internalIntrospectMemory }
        set { internalIntrospectMemory = newValue }
    }
}

typealias TestSentryCrashWrapper = TestSentryCrashReporter
#endif // !SDK_V10

final class TestSentryApplicationStateProvider: NSObject, SentryApplicationStateProvider {
    var isApplicationInForeground = true
}

final class TestSentryMemoryMetricsProvider: SentryMemoryMetricsProvider {
    var freeMemorySize: UInt64 = 0
    var appMemorySize: UInt64 = 0
    var usableMemorySize: UInt64 = 0
    var totalMemorySize: UInt64 = 0
}

final class TestSentryScopeContextEnricher: NSObject, SentryScopeContextEnricher {
    var enrichScopeCalled = false

    func enrichScope(_ scope: Scope) {
        enrichScopeCalled = true
    }
}
